esp/logging/logginglib/modularlogagent.cpp

```diff
@@ -0,0 +1,1526 @@
+    {
+        m_inheritTracePriorityLimit = true;
+    }
+    else
+    {
+        m_inheritTracePriorityLimit = false;
+        m_tracePriorityLimit = LogMsgDetail(tracePriorityLimit);
+    }
+
+    // Other initialization
+    m_disabled = configuration.getPropBool(propDisabled, m_disabled);
+
+    return true;
+}
+
+bool CModule::isEnabled() const
+{
+    return !m_disabled;
+}
+
+StringBuffer& CModule::toString(StringBuffer& str) const
+{
+    str.appendf("%s(", traceId());
+    appendProperties(str);
+    str.append(")");
+    return str;
+}
+
+bool CModule::appendProperties(StringBuffer& str) const
+{
+    str.appendf("disabled: %d; trace-priority-limit: %d", !isEnabled(), (m_inheritTracePriorityLimit ? -1 : int(m_tracePriorityLimit)));
+    return true;
+}
+
+//////////////////// CDelegatingAgent ////////////////////
+
+bool CDelegatingAgent::configure(const IPTree& configuration, const CModuleFactory& factory)
+{
+    bool result = Base::configure(configuration, factory);
+
+    if (!configureUpdateLog(configuration, factory))
+        result = false;
+
+    if (result && !hasService(LGSTGetTransactionSeed) && !hasService(LGSTGetTransactionID) && !hasService(LGSTUpdateLOG))
+        uwarnlog(Highest, "no available log agent services");
+
+    return result;
+}
+
+bool CDelegatingAgent::getTransactionSeed(IEspGetTransactionSeedRequest& request, IEspGetTransactionSeedResponse& response)
+{
+    return false;
+}
+
+void CDelegatingAgent::getTransactionID(StringAttrMapping* fields, StringBuffer& id)
+{
+}
+
+inline bool delegateUpdateLog(IUpdateLog& delegate, const char* content, IEspUpdateLogResponse& response)
+{
+    delegate.updateLog(content, response);
+    return (response.getStatusCode() == 0);
+}
+
+void CDelegatingAgent::updateLog(IEspUpdateLogRequestWrap& request, IEspUpdateLogResponse& response)
+{
+    bool result = false;
+    if (!m_updateLog)
+        ierrlog(Major, "UpdateLog service module does not exist");
+    else if (!m_updateLog->isEnabled())
+        m_updateLog->updateLog(nullptr, response);
+    else
+    {
+        const char* updateLogRequest = request.getUpdateLogRequest();
+        if (!isEmptyString(updateLogRequest))
+        {
+            result = delegateUpdateLog(*m_updateLog, updateLogRequest, response);
+        }
+        else
+        {
+            IPTree* ulrTree = request.getLogRequestTree();
+            if (ulrTree)
+            {
+                StringBuffer xml;
+                toXML(ulrTree, xml);
+                result = delegateUpdateLog(*m_updateLog, xml, response);
+            }
+            else
+            {
+                m_updateLog->ierrlog(Major, "requires XML markup; none given");
+            }
+        }
+    }
+
+    if (result && response.getStatusCode() != 0)
+    {
+        response.setStatusCode(0);
+    }
+    else if (!result && response.getStatusCode() == 0)
+    {
+        response.setStatusCode(-1);
+    }
```

Comment:
Yes. The response was passed by reference to the delegate module. I do not want to overwrite a potentially more descriptive status code set by the delegate.

Reply:
The delegateUpdateLog calls will remain above. The if condition does not need to call getStatusCode because result will only be true if getStatusCode returns 0. The else if condition does not need to test result, but does need to distinguish failures set by the delegate from local failures.